types/next/router.d.ts

```diff
@@ -78,14 +80,14 @@ export interface RouterProps {
     };
 }
 
-export interface SingletonRouter extends RouterProps {
-    router: RouterProps | null;
+export interface SingletonRouter<Q = DefaultQuery> extends RouterProps<Q> {
+    router: RouterProps<Q> | null;
     readyCallbacks: RouterCallback[];
     ready(cb: RouterCallback): void;
 }
 
-export interface WithRouterProps {
-    router: SingletonRouter;
+export interface WithRouterProps<Q = DefaultQuery> {
+    router: SingletonRouter<Q>;
 }
 
 export function withRouter<T extends {}>(
```

Comment:
To fix the single-use generic issue on `dtslint` you may be able to replace `withRouter` with the following:

```ts
export const withRouter: WithRouterFn;
```

and including `WithRouterFn` as a separate, exportable type:

```ts
export type WithRouterFn = <T extends {}, Q = DefaultQuery>(Component: React.ComponentType<T & WithRouterProps<Q>>): React.ComponentType<T>;
```

Reply:
Thank you for the idea! I tried it out and the issue persisted. I can understand the error, however, this is a weird use case where we have it present for type inference in (maybe) rare use cases. I opted for overriding the linter and would be more than happy to go back and poke more.